orttraining/orttraining/training_api/module.cc

```diff
@@ -2,12 +2,137 @@
+  ORT_ENFORCE(data_.IsAllocated());
+  ORT_ENFORCE(requires_grad_);
+  gradient_name_ = gradient_name;
+  ORT_ENFORCE(OrtValueLike(sess_state, data_, gradient_).IsOK());
+  return Status::OK();
+}
+
+Status Parameter::ResetGrad() {
+  //TODO: make use of lazy_reset_grad input instead 
+  Tensor* p_tensor = gradient_.GetMutable<Tensor>();
+  const auto& device = p_tensor->Location().device;
+  if (device.Type() == OrtDevice::CPU) {
+    memset(p_tensor->MutableDataRaw(), 0, p_tensor->SizeInBytes());
+  }
+#if defined(USE_CUDA) || defined(USE_ROCM)
+  else if (device.Type() == OrtDevice::GPU) {
+    ORT_NOT_IMPLEMENTED("Not implemented.");
+  }
+#endif
+  else {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Unknown device type ", device.Type(), " for param:", name_);
+  }
+  return Status::OK();
+}
+Module::Module(const std::string& train_model_path_or_bytes,
+               std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters,
+               const std::optional<std::string>& eval_model_path_or_bytes) {
+  // create value copy as same params are passed to Optimizer constructor
+  parameters_ = parameters;
+
+  auto so = onnxruntime::SessionOptions();
+  std::unique_ptr<Environment> env;
+  ORT_THROW_IF_ERROR(Environment::Create(nullptr, env));
+  train_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+  ORT_THROW_IF_ERROR(train_sess_->Load(train_model_path_or_bytes));
+  ORT_THROW_IF_ERROR(train_sess_->Initialize());
+  if (eval_model_path_or_bytes.has_value()) {
+    std::shared_ptr<onnxruntime::Model> eval_model;
+    ORT_THROW_IF_ERROR(onnxruntime::Model::Load(eval_model_path_or_bytes.value(), eval_model,
+                                                nullptr, env->GetLoggingManager()->DefaultLogger()));
+    GetGraphInputOutputNames(eval_model->MainGraph(), eval_input_names_, eval_output_names_);
+    // TODO:: do validation on eval inputs and outputs: eg order of user inputs, weights
+    eval_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+    ORT_THROW_IF_ERROR(eval_sess_->Load(eval_model_path_or_bytes.value()));
+    ORT_THROW_IF_ERROR(eval_sess_->Initialize());
+  }
+  auto& train_sess_state = train_sess_->GetSessionState();
+
+  std::shared_ptr<onnxruntime::Model> train_model;
+  ORT_THROW_IF_ERROR(onnxruntime::Model::Load(train_model_path_or_bytes, train_model,
+                                              nullptr, env->GetLoggingManager()->DefaultLogger()));
+  GetGraphInputOutputNames(train_model->MainGraph(), train_input_names_, train_output_names_);
+
+  std::vector<std::string> param_input_names, grad_input_names, user_input_names;
+  std::string param_name;
+  for (auto input_name : train_input_names_) {
+    auto it = parameters_.find(input_name);
+    if (it != parameters_.end()) {
+      param_input_names.emplace_back(input_name);
+      weights_.push_back(it->second->Data());
+    } else if (GetParamNameFromGradient(input_name, param_name)) {
```

Comment:
I assume train_input_names_ includes name of gradient accumulation buffer, do you mean that one?

Reply:
yes, finding the param name for the gradient name. Hopefully we can do away with this string matching by encoding this info in the checkpoint in some way